File: 2021/09/06/find_mode.py
"""
https://leetcode.com/problems/find-mode-in-binary-search-tree/
Author: Andrew Jarombek
Date: 9/6/2021
"""

from typing import Optional, List


class TreeNode:
    def __init__(self, x, left=None, right=None):
        self.val = x
        self.left = left
        self.right = right


class Solution:
    mode = 0
    modes = []
    prior_value = None
    consecutive_values = 0

    def find_mode(self, root: Optional[TreeNode]) -> List[int]:
        self.search(root)

        if self.consecutive_values > self.mode:
            self.modes = [self.prior_value]
        elif self.consecutive_values == self.mode:
            self.modes.append(self.prior_value)

        return self.modes

    def search(self, root: Optional[TreeNode]) -> None:
        if root.left is not None:
            self.search(root.left)

        if root.val == self.prior_value:
            self.consecutive_values += 1
        else:
            if self.consecutive_values > self.mode:
                self.mode = self.consecutive_values
                self.modes = [self.prior_value]
            elif self.consecutive_values == self.mode:
                self.modes.append(self.prior_value)

            self.prior_value = root.val
            self.consecutive_values = 1

        if root.right is not None:
            self.search(root.right)


if __name__ == '__main__':
    solution = Solution()
    assert solution.find_mode(TreeNode(1, None, TreeNode(2, TreeNode(2), None))) == [2]

File: 2022/09/07/findMode.js
/**
 * https://leetcode.com/problems/find-mode-in-binary-search-tree/
 * @author Andrew Jarombek
 * @since 9/7/2022
 */

function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

const findMode = function(root) {
    const result = dfs(root, {mode: 0, count: 0, prev: 0, values: []});
    return result.values;
};

const dfs = function(node, args) {
    if (node.left) {
        args = dfs(node.left, args);
    }

    let {mode, count, prev, values} = args;

    if (node.val === prev) {
        count++;
    } else {
        prev = node.val;
        count = 1;
    }

    if (count > mode) {
        values = [prev];
        mode = count;
    } else if (count === mode) {
        values.push(prev);
    }

    args = {mode, count, prev, values};

    if (node.right) {
        args = dfs(node.right, args);
    }

    return args;
}

let result = findMode(new TreeNode(1, null, new TreeNode(2)));
let expected = [1, 2]
console.assert(expected.every((value, index) => value === result[index]));

result = findMode(new TreeNode(1, null, new TreeNode(2, new TreeNode(2), null)));
expected = [2]
console.assert(expected.every((value, index) => value === result[index]));
